### ISSUE TITLE
Binary operator of aggregation fails.

### ISSUE THREAD

**Issue by useredsa** [VictoriaMetrics/VictoriaMetrics#5604]
### Describe the bug

The following code fails and does not output the sum. Swapping the order of the arguments to the aggregation function solves the issue (see commented line).

```
with(
  args = {namespace="argocd"},
  args_not_matching = {args, namespace="namespace cannot be equal to this"},

  metric = avg by(namespace, container) (go_gc_duration_seconds_count{args}[1i]),
  neg_metric = - avg by(namespace, container) (go_gc_duration_seconds_count{args}[1i]),
  metric_not_matching = avg by(namespace, container) (go_gc_duration_seconds_count{args_not_matching}[1i]),
  # metric_sum = sum by (namespace, container)(metric, metric_not_matching), # This works
  metric_sum = avg by (namespace, container)(metric_not_matching, metric), # This doesn't (you only get series in first argument)
)

(
  alias(metric_sum, "pos"),
  alias(neg_metric, "neg"),
  alias(metric_sum + neg_metric, "sum"),
)
```

The bug is reproducible with any series or aggregation function, not just `avg by`, and any binary operator, like `and`, not just `+`.

Also, if we change the "not matching arguments" so that they match something but not all series, then only the series present in that variable are summed. My understanding is that the optimizer is trying to avoid work and checks the labels that would arise from each side of the binary operator to not compute them if they'd be dropped. And for the aggregation functions, the optimizer is bugged and only takes labels from the first argument. That's why changing the arguments' order solves the issue.

### To Reproduce

[Bug reproduced in playground.](https://play.victoriametrics.com/select/accounting/1/6a716b0f-38bc-4856-90ce-448fd713e3fe/prometheus/graph/#/?g0.expr=with%28%0A++args+%3D+%7Bnamespace%3D%22argocd%22%7D%2C%0A++args_not_matching+%3D+%7Bargs%2C+namespace%3D%22namespace+cannot+be+equal+to+this%22%7D%2C%0A%0A++metric+%3D+avg+by%28namespace%2C+container%29+%28go_gc_duration_seconds_count%7Bargs%7D%5B1i%5D%29%2C%0A++neg_metric+%3D+-+avg+by%28namespace%2C+container%29+%28go_gc_duration_seconds_count%7Bargs%7D%5B1i%5D%29%2C%0A++metric_not_matching+%3D+avg+by%28namespace%2C+container%29+%28go_gc_duration_seconds_count%7Bargs_not_matching%7D%5B1i%5D%29%2C%0A++%23+metric_sum+%3D+sum+by+%28namespace%2C+container%29%28metric%2C+metric_not_matching%29%2C+%23+This+works%0A++metric_sum+%3D+avg+by+%28namespace%2C+container%29%28metric_not_matching%2C+metric%29%2C+%23+This+doesn%27t+%28you+only+get+series+in+first+argument%29%0A%29%0A%0A%28%0A++alias%28metric_sum%2C+%22pos%22%29%2C%0A++alias%28neg_metric%2C+%22neg%22%29%2C%0A++alias%28metric_sum+%2B+neg_metric%2C+%22sum%22%29%2C%0A%29&g0.range_input=30m&g0.end_input=2024-01-11T10%3A42%3A20&g0.tab=1&g0.relative_time=last_30_minutes)

### Version

I encountered the error in 1.93.6, which is LTS, but the bug is also present on the latest 1.96.0.

**Comment by useredsa** [comment]
Also, the bug dissapears if you realias the series before the binary operator, like
```
alias(metric_sum, "s1") + alias(neg_metric, "s2")
```

**Comment by useredsa** [comment]
Any updates in this? I think it's a fairly important bug that could affect production code. The fact that some series are dropped randomly, unexpectedly and without any warning makes it very dangerous.

**Comment by valyala** [comment]
@useredsa , thanks for filing the bug with the detailed description! It should be fixed in the commit 61524ad87ba922d8aa38e7b81d3dae590b852a72 . This commit will be included in the next release. In the mean time you can build VictoriaMetrics from this commit according to [these docs](https://docs.victoriametrics.com/#how-to-build-from-sources) and verify whether the issue is fixed.

If you run cluster version of VictoriaMetrics, then you can build it from the commit ebd0f1b72e2487639808c7268699fac8ac30636c according to [these docs](https://docs.victoriametrics.com/cluster-victoriametrics/#building-from-sources) and verify whether it works as intended.

You can also verify whether [VictoriaMetrics playground](https://play.victoriametrics.com/) is free from the bug.

**Comment by useredsa** [comment]
Thank you. It looks fixed. I checked using the playground.

Best regards.

**Comment by valyala** [comment]
FYI, the bugfix for this issue has been included in [VictoriaMetrics v1.98.0](https://docs.victoriametrics.com/CHANGELOG.html).

**Comment by valyala** [comment]
FYI, the bugfix has been included also in [v1.97.2 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.97.2).

**Comment by valyala** [comment]
FYI, the bugfix has been included also in [v1.93.12 LTS release](https://github.com/VictoriaMetrics/VictoriaMetrics/releases/tag/v1.93.12).
